### PR TITLE
gc_worker: limit the write flow of GC

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

## What have you changed? (mandatory)
* Use `IOLimiter` to limit the write flow of GC.
* Make `GC_BATCH_SIZE` configurable.
* Don't fill cache when does GC.

## What are the type of changes? (mandatory)
- Improvement (a change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
Unit tests.

## Does this PR affect documentation (docs) or release note? (mandatory)
No

### CODE REVIEW

File: src/storage/config.rs [hunk 1]
```diff
@@ -109,3 +109,51 @@ impl BlockCacheConfig {
         Some(Cache::new_lru_cache(cache_opts))
     }
 }
+
+pub const DEFAULT_GC_RATIO_THRESHOLD: f64 = 1.1;
+pub const DEFAULT_GC_BATCH_SIZE: usize = 512;
+const DEFAULT_GC_MAX_WRITE_BYTES_PER_SEC: u64 = 5 * 1024 * 1024;
```
Comment (lonng) [inline src/storage/config.rs]: Prefer to `5 * tikv_util::config::MB`.

File: etc/config-template.toml [hunk 2]
```diff
@@ -158,6 +158,18 @@
...
+## The threshold of redundant ratio in a region to trigger GC.
+## The default value 1.1 should work for most cases. Set it to a value less than 1.0 to always trigger GC.
+# ratio-threshold = 1.1
+
+## The number of keys to GC in one batch.
+# batch-size = 512
+
+## Max bytes that GC worker can write to rocksdb in one second.
+## If it is set to 0, there is no limit.
+# max-write-bytes-per-sec = "5MB"
```
Comment (MyonKeminta) [inline etc/config-template.toml]: I prefer if this can be set without restating TiKV

File: etc/config-template.toml [hunk 3]
```diff
@@ -158,6 +158,18 @@
 ## Otherwise, the OOM problem might occur in TiKV.
 # capacity = "1GB"
 
+[storage.gc]
+## The threshold of redundant ratio in a region to trigger GC.
+## The default value 1.1 should work for most cases. Set it to a value less than 1.0 to always trigger GC.
+# ratio-threshold = 1.1
+
+## The number of keys to GC in one batch.
+# batch-size = 512
```
Comment (MyonKeminta) [inline etc/config-template.toml]: Consider try to imply in the name that the batch size is by key, not bytes.
For example,  batch-keys or max-keys-in-batch or something

File: src/storage/gc_worker.rs [hunk 4]
```diff
@@ -251,8 +259,12 @@ impl<E: Engine> GCRunner<E> {
         }
         self.stats.add_statistics(&txn.take_statistics());
 
+        let write_size = txn.write_size();
         let modifies = txn.into_modifies();
         if !modifies.is_empty() {
+            if let Some(limiter) = &self.limiter {
+                limiter.request(write_size as i64);
```
Comment (lonng) [inline src/storage/gc_worker.rs]: The gc_worker will be blocked here if the limit exceeded, is it expected?

File: src/storage/config.rs [hunk 5]
```diff
@@ -31,27 +30,27 @@ const DEFAULT_SCHED_PENDING_WRITE_MB: u64 = 100;
 #[serde(rename_all = "kebab-case")]
 pub struct Config {
     pub data_dir: String,
-    pub gc_ratio_threshold: f64,
```
Comment (MyonKeminta) [inline src/storage/config.rs]: This might break backward compatibility.

File: src/storage/gc_worker.rs (hunk 4 above)
Comment (youjiali1995) [reply src/storage/gc_worker.rs]: yes

Comment (MyonKeminta) [reply src/storage/gc_worker.rs]: I think yes. We may avoid the blocking when refactoring it with futures.

File: src/storage/config.rs (hunk 5 above)
Comment (MyonKeminta) [reply src/storage/config.rs]: Please consider remove it from the config template but keep it in code here.

File: etc/config-template.toml (hunk 2 above)
Comment (youjiali1995) [reply etc/config-template.toml]: Of course it can be. Maybe I can do it in a new PR.

Comment (zhangjinpeng1987) [reply etc/config-template.toml]: Do we need a `max-write-keys-per-sec`?

**Comment by youjiali1995** [comment]
@zhangjinpeng1987 @AndreMouche @MyonKeminta PTAL

File: src/storage/config.rs [hunk 6]
```diff
@@ -109,3 +112,51 @@ impl BlockCacheConfig {
         Some(Cache::new_lru_cache(cache_opts))
     }
 }
+
+pub const DEFAULT_GC_RATIO_THRESHOLD: f64 = 1.1;
+pub const DEFAULT_GC_BATCH_KEYS: usize = 512;
+const DEFAULT_GC_MAX_WRITE_BYTES_PER_SEC: u64 = 5 * MB;
```
Comment (zhangjinpeng1987) [inline src/storage/config.rs]: I think the default value should no limit?

Comment (youjiali1995) [reply src/storage/config.rs]: Okay

File: etc/config-template.toml (hunk 2 above)
Comment (youjiali1995) [reply etc/config-template.toml]: No. I think `max-write-bytes-per-sec` is enough.

File: etc/config-template.toml [hunk 7]
```diff
@@ -158,6 +158,18 @@
 ## Otherwise, the OOM problem might occur in TiKV.
 # capacity = "1GB"
 
+[storage.gc]
```
Comment (siddontang) [inline etc/config-template.toml]: I know @MyonKeminta wants to separate GC from storage module, so can we use `[gc]` section direclty？

**Comment by siddontang** [comment]
em, can we use one IOLimiter for snapshot, gc, and compaction?

@zhangjinpeng1987

**Comment by MyonKeminta** [comment]
Despite of how it helps on our current problem, will we continue and merge this PR?

**Comment by zhangjinpeng1987** [comment]
> em, can we use one IOLimiter for snapshot, gc, and compaction?

I don't think so, for snapshot we mainly limit the IO usage, but for GC mainly limit the front ground write flow to the LSM tree.

**Comment by MyonKeminta** [comment]
Can we test how this PR affects .99 and MAX latency during GC?

**Comment by youjiali1995** [comment]
> Can we test how this PR affects .99 and MAX latency during GC?

I don't have time to test it this week.😥 And if https://github.com/tikv/tikv/pull/5333 is merged, this PR is useless.

**Comment by youjiali1995** [comment]
Since https://github.com/tikv/tikv/pull/5333 achieved much improvement, close this PR.

**Comment by youjiali1995** [comment]
@MyonKeminta @zhangjinpeng1987 @AndreMouche PTAL

File: etc/config-template.toml [hunk 8]
```diff
@@ -747,3 +747,11 @@
...
 ## release locks that pessimistic transactions are waiting for(0 to disable).
 # wake-up-delay-duration = 1
+
+[gc]
+## The number of keys to GC in one batch.
+# batch-keys = 512
+
+## Max bytes that GC worker can write to rocksdb in one second.
+## If it is set to 0, there is no limit.
+# max-write-bytes-per-sec = "0"
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply etc/config-template.toml]: I think it's a hidden configuration which shouldn't config by users. It's an idiom that hidden configurations are not listed in `config-template.toml`. It's the reason why many configurations are missing.😂

**Comment by AndreMouche** [comment]
Please cherry-pick manually @youjiali1995

**Pull request merged**